### PR TITLE
Reliable message delivery design proposal

### PR DESCRIPTION
**Pull request by rohitsardesai83**
**What type of PR is this?**
/kind design

**What this PR does / why we need it**:
Enable reliable message delivery from cloud to edge.

**Which issue(s) this PR fixes**:

xref #725 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by kadisi** [comment]
hi @rohitsardesai83,  

even though  any of the cloud components crash, in-flight events will be lost. Messages stored by the Event Queue in-memory would be destroyed. but when edgecontroller restart, event queue will get all event, because we use kubernetes List and watch mechanism .We can get the current full amount of data from kubernetes when edgecontroller restart.

In order to reduce invalid data transmission, should we be concerned about the problem of data duplication ? for example.

there are tow event which are same pod update in event queue .  But at this time they have not been sent to edge node, we just need send the last event.

another example.

there are tow event which are same pod in event queue, the first event is update pod, the second is delete pod, But at this time they have not been sent to edge node , we just need send the second event.

**Comment by rohitsardesai83** [comment]
> hi @rohitsardesai83,
> 
> even though any of the cloud components crash, in-flight events will be lost. Messages stored by the Event Queue in-memory would be destroyed. but when edgecontroller restart, event queue will get all event, because we use kubernetes List and watch mechanism .We can get the current full amount of data from kubernetes when edgecontroller restart.
> 
> In order to reduce invalid data transmission, should we be concerned about the problem of data duplication ? for example.
> 
> there are two event which are same pod update in event queue . But at this time they have not been sent to edge node, we just need send the last event.
> 
> another example.
> 
> there are two event which are same pod in event queue, the first event is update pod, the second is delete pod, But at this time they have not been sent to edge node , we just need send the second event.

Yes, we need to think about how to handle duplicate events. 
Kubernetes did something similar : https://github.com/kubernetes/kubernetes/issues/4073.

File: docs/proposals/reliable-message-delivery.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+There is no reliable mechanism to exchange messages between the cloud and the edge nodes.
+The current approach uses a fire-and-forget approach where messages are sent to edge nodes and there is no way to confirm if
+those messages were actually received by the edge node. Unstable networks can result in frequent disconnection of edge nodes and
+this can result in loss of messages sent to edge nodes which can’t be temporarily reached. This proposal addresses this problem
+thus ensuring reliable message delivery.
+
+### Goals
+
+- Provide a reliable message delivery mechanism between cloud and edge.
```
Comment (sids-b) [inline docs/proposals/reliable-message-delivery.md]: from cloud to edge ?

**Comment by sids-b** [comment]
ping @qizha @kevin-wangzefeng @fisherxu for review

**Comment by fisherxu** [comment]
Dumplicate with #1343 , close this first.
/close

**Pull request closed without merging**